### PR TITLE
Mandatory fields for Assign

### PR DESCRIPTION
**Pull request by nickdrozd**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |
| ✓   | :hammer: Refactoring   |

## Description

This PR re-types the required fields of the `Assign` node so that they are non-optional, as described in https://github.com/PyCQA/astroid/issues/2060.

Two existing type errors are thereby fixed, as reported by Mypy. Here is one of them:

```python
    @cached_property
    def _assign_nodes_in_scope(self) -> list[nodes.Assign]:
        return [self, *self.value._assign_nodes_in_scope]
```

If `Assign.value` can be `None`, then this could fail. But if the field is required, there is no problem.

### CODE REVIEW

**Review by Pierre-Sassoulas — COMMENTED** [review]
Shouldn't we change the constructor for ``_base_nodes.AssignTypeNode, _base_nodes.Statement`` too ?

**Comment by DanielNoord** [comment]
@Pierre-Sassoulas Are we okay with merging breaking changes into `main`? I'd be in favour, but want to double check.

**Comment by Pierre-Sassoulas** [comment]
Let's consider that if it's not breaking pylint then it's not a change that require a major version bump even if technically it's a breaking change. Especially if it's typing, as we never advertised astroid as typed.

**Comment by nickdrozd** [comment]
> Shouldn't we change the constructor for `_base_nodes.AssignTypeNode, _base_nodes.Statement` too ?

It looks like those don't have their own constructors. They all just delegate to `NodeNG`.

**Review by Pierre-Sassoulas — COMMENTED** [review]
I'd rather have the overridden constructor requiring 'targets' and remove the post init function altogether. But for that to work we would need to refactor the tree rebuilder. I suppose if we instantiate the leafs first and then recurse instead of what we're currently doing it's possible (?). What do you think @DanielNoord ?

**Comment by DanielNoord** [comment]
> I'd rather have the overridden constructor requiring 'targets' and remove the post init function altogether. But for that to work we would need to refactor the tree rebuilder. I suppose if we instantiate the leafs first and then recurse instead of what we're currently doing it's possible (?). What do you think @DanielNoord ?

I mean, it's worth a shot but it would be a massive rewrite.

I would be okay with reviewing a complete rewrite and contributing but don't have the time to do to brunt of the work..

File: astroid/nodes/node_classes.py [hunk 1]
```diff
@@ -1202,62 +1202,25 @@ class Assign(_base_nodes.AssignTypeNode, _base_nodes.Statement):
...
-            lineno=lineno,
-            col_offset=col_offset,
-            end_lineno=end_lineno,
-            end_col_offset=end_col_offset,
-            parent=parent,
-        )
+    _astroid_fields = ("targets", "value")
+    _other_other_fields = ("type_annotation",)
 
     def postinit(
```
Comment (DanielNoord) [inline astroid/nodes/node_classes.py:1217]: https://github.com/pylint-dev/astroid/blob/345b91cd6d29318ac34352699ff75a0132ccda15/astroid/rebuilder.py#L808

Could you add asserts here that the typing we add here is correct?

And could you add a changelog entry? For the rest this LGTM!

**Review by Pierre-Sassoulas — APPROVED** [review]
I don't see myself doing a massive rewrite (when there are a lot of massive rewrite we could do) so let's merge, it's an upgrade in the current state of the code.

**Comment by DanielNoord** [comment]
> I don't see myself doing a massive rewrite (when there are a lot of massive rewrite we could do) so let's merge, it's an upgrade in the current state of the code.

Agreed! But let's at least make sure that all the callsites of these functions, so everywhere we do `Assign(...` are in line with what we merge. Otherwise we will start wondering in 3 months time if we forget something and the new typing is incorrect as well :sweat_smile:

**Comment by nickdrozd** [comment]
What about the rest of the node types? Many of the rest need a similar fix. (Some require more involved changes.) Should that be done here, or separate PRs? One PR per node type, or all at once?

**Comment by DanielNoord** [comment]
Let's do separate PRs! I think I can check these fairly quickly but focusing on one node at a time makes it much easier. You can tag me for review (although please don't open 10 at once 😄 )

**Review by DanielNoord — APPROVED** [review]

**Review by Pierre-Sassoulas — APPROVED** [review]

**Pull request merged**